Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -178,35 +202,50 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
         return;
     }
 
-    PAL::CMBufferQueueEnqueue(m_compressedSampleQueue.get(), cmSampleBuffer);
-    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }] {
-        if (RefPtr protectedThis = weakThis.get())
+    ++m_framesBeingDecoded;
+    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, sample = RetainPtr { cmSampleBuffer }, flushId = m_flushId.load()]() mutable {
+        if (RefPtr protectedThis = weakThis.get()) {
+            assertIsCurrent(protectedThis->m_workQueue.get());
+
+            if (flushId != protectedThis->m_flushId) {
```

Comment:
we don't want any of the frames to be decoded, nor those in the process of being decoded to be added to the queue. It also prevents blocking the main thread while we wait for a decode to complete.

Ultimately, yes, all that does similar in WebCoreDecompressionSession will be removed and we will only use `VideoMediaSampleRenderer` 

Having this in mind, it was easier to have the `VideoMediaSampleRenderer` manage the reordering of the decoding samples than have the WebCoreDecompressionSession do it. The WebCoreDecompressionSession doesn't have a timebase set , it decodes one sample at a time. So for the `WebCoreDecompressionSession` to do the re-ordering, it would need to keep the frames in the queue until the sliding window has been fully received adding latency.

As the VideoMediaSampleRenderer can simply pull the frame from the re-ordering queue only when it's actually time to use the frame, it guarantees that the frames are in the right order without latency.

I hinted at this in the commit log

Reply:
oh, m_flushId is increased whenever `flush()` is called